web/interceptor/access_control.go

```diff
@@ -0,0 +1,176 @@
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		req, ok := request.(requestID)
+		// The GetUserByCourse method sends a CourseUserRequest which has no IDs and needs a database query.
+		if !ok {
+			logger.Errorf("%s failed: message type %T does not implement IDFor interface", method, request)
+			return nil, ErrAccessDenied
+		}
+		claims, err := tm.GetClaims(ctx)
+		if err != nil {
+			logger.Errorf("AccessControl(%s): failed to get claims from request context: %v", method, err)
+			return handler(ctx, request)
+		}
+		for _, role := range accessRolesFor[method] {
+			switch role {
+			case none:
+				return handler(ctx, request)
+			case user:
+				if claims.SameUser(req) {
+					// Make sure the user is not updating own admin status.
+					if method == "UpdateUser" {
+						if req.(*qf.User).GetIsAdmin() && !claims.Admin {
+							logger.Errorf("AccessControl(%s): user %d attempted to change admin status from %v to %v",
+								method, claims.UserID, claims.Admin, req.(*qf.User).GetIsAdmin())
+							return nil, ErrAccessDenied
+						}
+					}
+					return handler(ctx, request)
+				}
+			case student:
+				// GetSubmissions is used to fetch individual and group submissions.
+				// For individual submissions needs an extra check for user ID in request.
+				if method == "GetSubmissions" && req.IDFor("group") == 0 {
+					if !claims.SameUser(req) {
+						logger.Errorf("AccessControl(%s): ID mismatch in claims (%s) and request (%s)",
+							method, claims.UserID, req.IDFor("user"))
+						return nil, ErrAccessDenied
+					}
+				}
+				if claims.HasCourseStatus(req, qf.Enrollment_STUDENT) {
+					return handler(ctx, request)
+				}
+			case group:
+				// Request for CreateGroup will not have ID yet, need to check
+				// if the user is in the group (unless teacher).
+				if method == "CreateGroup" {
+					notMember := !req.(*qf.Group).Contains(&qf.User{ID: claims.UserID})
+					notTeacher := !claims.HasCourseStatus(req, qf.Enrollment_TEACHER)
+					if notMember && notTeacher {
+						logger.Errorf("AccessControl(%s): user %d tried to create group while not teacher or group member", method, claims.UserID)
+						return nil, ErrAccessDenied
+					}
+					// Otherwise, create the group.
+					return handler(ctx, request)
+				}
+				groupID := req.IDFor("group")
+				for _, group := range claims.Groups {
+					if group == groupID {
+						return handler(ctx, request)
+					}
+				}
+			case teacher:
+				if method == "GetUserByCourse" {
+					if err := claims.IsCourseTeacher(tm.Database(), request.(*qf.CourseUserRequest)); err != nil {
+						logger.Errorf("AccessControl(%s): %v", method, err)
+						return nil, ErrAccessDenied
+					}
+					return handler(ctx, request)
+				}
+				if claims.HasCourseStatus(req, qf.Enrollment_TEACHER) {
+					return handler(ctx, request)
+				}
+			case courseAdmin:
+				if claims.Admin {
+					if method == "GetUserByCourse" {
```

Comment:
The `GetUserByCourse` method does not list the `courseAdmin` role.

Also, there doesn't seem to be any tests for the `courseAdmin` role, since this doesn't cause a test failure:
```go
case courseAdmin:
	return nil, ErrAccessDenied
```